File: 145/BinaryTreePostorderTraversal.scala
object BinaryTreePostorderTraversal {
  def postorderTraversal(root: TreeNode): List[Int] = {
    @annotation.tailrec
    def postorderTraversal(stack: List[TreeNode], res: List[Int]): List[Int] = stack match {
      case h :: t =>
        val r = h.value :: res
        if (h.left != null && h.right != null)
          postorderTraversal(h.left :: h.right :: t, r)
        else if (h.right != null)
          postorderTraversal(h.right :: t, r)
        else if (h.left != null)
          postorderTraversal(h.left :: t, r)
        else
          postorderTraversal(t, r)

      case _ => res.reverse
    }

    if (root == null) List.empty
    else postorderTraversal(root :: Nil, Nil)
  }

  class TreeNode(var _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }
}

File: 145/binary-tree-postorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    let ans = [];

    function dfs(root, ans) {
        if (!root) return;

        dfs(root.left, ans);
        dfs(root.right, ans);
        ans.push(root.val);
    }

    dfs(root, ans);
    return ans;
};